### DISCUSSION TITLE
How to find self-created ncmat data files by a convenient way, but not by absolute paths?

### DISCUSSION THREAD

**Discussion by YangNi20** [mctools/ncrystal#82] [Q&A] [answered]
Hi everyone，
I’m Ni Yang, a graduate student from Institute of High Energy Physics Chinese Academy of Sciences. I’m participating in the development of a Monte Carlo simulation system called Cinema(China Spallation Neutron Source Monte Carlo Simulation System), the NCrystal version of which is 3.0.0.
I would like to use the ncmat data files created by myself, while using the standard data library of NCrystal. These ncmat data files are put into another folder called ncmat.   
Now I can only use the absolute paths to find the data files in ncmat folder, which is cumbersome. Dose anyone know a more convenient way to find self-created ncmat data files?

Best Regards,
Ni

**Comment by tkittel** [comment, marked as answer]
Hi @YangNi20,

Thank you for your question - "Cinema" sounds like an interesting project and I am happy to hear you are using NCrystal :-)

Fortunately, there are several ways you can make NCrystal find your files:

From Python:
```python
import NCrystal as NC
NC.addCustomSearchDirectory("/path/to/your/files")
```

From C++:
```C++
#include "NCrystal/NCrystal.hh"
//...
NCrystal::DataSources::addCustomSearchDirectory("/path/to/your/files");
```

Or you can simply add your directory to the `NCRYSTAL_DATA_PATH` environment variable, for instance from a shell:
```sh
export NCRYSTAL_DATA_PATH="/path/to/your/files:$NCRYSTAL_DATA_PATH"
```

Finally, I should perhaps remind you that you can also register "files" in-memory rather than having to create physical files - that is particularly useful when you are autogenerating NCMAT data on the fly (see for instance the python/C++ functions `registerInMemoryFileData`).

Cheers,
Thomas

**Comment by tkittel** [comment]
@YangNi20 did the answer work for you? (a short reply or marking the answer as accepted would be nice).

**Reply by YangNi20** [reply]
Hi Thomas,
Thank you for your kind help and I’m very sorry for the late reply. 
I have tried your last suggestion which seems more suitable for Cinema, but it didn’t work well.
If my ncmat directory isn’t added to the` NCRYSTAL_DATADIR `environment variable, all the ncmat files can be found with the command `ncrystal_inspectfile -b` , including self-created ncmat files and the standard data. Here is the result:
![QQ图片20220606204726](https://user-images.githubusercontent.com/49681051/172166179-bc01ca14-8450-4771-9b45-ff3da6e5f56c.png)
After I add my ncmat directory, it’s confusing that only the standard data library can not be found with the command `ncrystal_inspectfile -b` . Here is the result:
![QQ图片20220606205209](https://user-images.githubusercontent.com/49681051/172166288-61db1de3-ed03-49b9-9cda-dc8e94388115.png)
Do you know how to solve this problem? I would be very glad if you could help me.
Best Regards,
Ni

**Comment by tkittel** [comment]
Hi @YangNi20 . You used the wrong variable name. The `NCRYSTAL_DATADIR` variable is used to override the location of the NCrystal standard library, so you should normally not touch this variable at all (it is only needed if the built-time installation directory needs to be overriden at run-time). It is also not a path-like variable (i.e. it does **not** support multiple entries separated by `:`). The variable I suggested to you is named `NCRYSTAL_DATA_PATH`.

Btw., the results from the `relpath` factory simply shows files lying around in your current working directory or in one of the immediate sub-directories.

**Reply by YangNi20** [reply]
Hi Thomas,
Thank you for your informative reply! According to your advice, my problem is perfectly solved. After adding the directory to the `NCRYSTAL_DATA_PATH` environment variable, I can use self-created ncmat files by not directing absolute paths.

Best Regards,
Ni
